torch/csrc/jit/graph_executor.cpp

```diff
@@ -367,21 +367,22 @@ struct DifferentiableGraphOp {
  private:
   friend GraphExecutor* detail::getGradExecutor(Operation& op);
 
-  void detach(at::Tensor& t) const {
-    if (t.defined()) {
-      t = autograd::as_variable_ref(t).detach();
+  at::Tensor detach(at::Tensor t) const {
+    if (!t.defined()) {
+      return t;
     }
+    return autograd::as_variable_ref(std::move(t)).detach();
   }
 
   void detach(IValue& v) const {
     if (v.isTensor()) {
       auto t = std::move(v).toTensor();
-      detach(t);
+      t = detach(t);
```

Comment:
This is a perf regression. There are more references counts and move constructors happening here than before. This is one reason I am wary about this 'future proof' abstractions. They obscure what is actually going on an make performance optimization more opaque.

Reply:
Ah, forgot a `std::move` here. Thanks. Adding that, it's not a perf regression anymore. Move constructors of Tensors are for free, only copies bump the refcount..

Reply:
after adding another `std::move`, now it's even one refcount bump less than the original.